### ISSUE TITLE
Code using twisted.names.client.Resolver is hard to unit test

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3908]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3908|
|Type|enhancement|
|Created|2009-07-09 18:45:00Z|
|Branch|https://github.com/twisted/twisted/tree/testable-resolver-3908|

`Resolver` uses `IReactorUDP` and `IReactorTime`, but also uses the global reactor and does not allow this to be overridden (using public APIs).  This means any code using `Resolver` will probably do actual UDP traffic and use the actual system clock.  This makes it hard to test.

Parameterizing the reactor would all alternate UDP/time implementations to be supplied.  This would be nice.  Alternatively, Twisted Names could provide a verified fake of the `Resolver` interface, and tests could use that instead of the actual `Resolver`.  This fake should have easily controlled behavior, including error cases.

Attachments:

* [testable-examples-6362-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f02/f0296ccc1729c39727da009b09867730fd00b8f4/bfdd713b770dab0bf2b21435b90a6fbf1ddcde8a.patch) (2238 bytes) - added by rwall on 2013-03-09 20:58:16Z - This is the patch where I pass the reactor through

<details><summary>Searchable metadata</summary>

```
trac-id__3908 3908
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_testable_resolver_3908 branches/testable-resolver-3908
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1247165100000000 1247165100000000
changetime__1363647054000000 1363647054000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

I'm trying to test the [docs/names/examples/testdns.py](docs/names/examples/testdns.py) in [#6362](https://github.com/twisted/twisted/issues/6362)

I can pass a fake reactor to client.Resolver but it isn't passed to dns.DNSDatagramProtocol, so I get unclean reactor errors in the tests like exarkun predicted.

I also found an existing MemoryReactor in [twisted/names/test/test_rootresolve.py](twisted/names/test/test_rootresolve.py) which does implement IReactorUDP. So maybe that can be merged with the proto_helpers.MemoryReactor as part of this ticket.

See attachement:testable-examples-6362-1.patch for an example of what I am trying to do.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj commented|
|-|-|

(In [37479]) Branching to 'testable-resolver-3908'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for review in log:branches/testable-resolver-3908

 * [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/testable-resolver-3908)
 * Added a test that client.Resolver uses its parameterized reactor for delayed calls during resolv.conf checking and in udp query cancellation.
 * Pass the parameterized reactor to the DNSDatagramProtocol to prevent it loading the global reactor.


The ticket implies that client.Resolver doesn't accept a reactor
parameter. It does, but it does not pass that reactor to the
DNSDatagramProtocol used in the queries.

Therefore the global reactor is used by DNSDatagramProtocol to
schedule a query cancellation function.

This triggers dirty reactor warnings in trial when you attempt to test
client.Resolver (or in my case, the examples that use it [[#6362](https://github.com/twisted/twisted/issues/6362)](https://github.com/twisted/twisted/issues/6362).)

I've added some assertions to the new test method, but they seem quite
tightly coupled to the implementation of client.Resolver. I wondered if
it was sufficient to simply write a test which should not trigger
dirty reactor warnings.

I haven't attempted to create a verified fake of IResolver as
suggested in the ticket description. I don't need it for #6362 and
it seems like a big enough task to split into a separate ticket. I'll
create a ticket for that if the reviewer agrees.

I also didn't attempt to merge t.n.test.test_rootresolve.MemoryReactor into
t.test.proto_helpers.MemoryReactor. Would that be a good thing to do? If so,
I'll create another ticket for that too.

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince set owner to @wallrj|
|-|-|

 - I think it would be better to test `_connectedProtocol` directly, rather some public thing that eventually calls it. Then you should be able to write a better assertion.
 - That probably wants to use `MemoryReactor` from `test_rootresolver`. That should be merged to the core `MemoryReactor` (or better, composed with, using exarkun's idea of composable reactor bits). But the latter would be a separate ticket.
 - In any case, we separately want a verified fake resolver. But that should be done in a separate ticket.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj removed owner|
|-|-|

Ready for another review in log:branches/testable-resolver-3908:

 * [Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/testable-resolver-3908)


Replying to [tom.prince](#note_5):
>  - I think it would be better to test `_connectedProtocol` directly, rather some public thing that eventually calls it. Then you should be able to write a better assertion.

Done. Much simpler.

>  - That probably wants to use `MemoryReactor` from `test_rootresolver`. That should be merged to the core `MemoryReactor` (or better, composed with, using exarkun's idea of composable reactor bits). But the latter would be a separate ticket.

It was already using that version of MemoryReactor.

>  - In any case, we separately want a verified fake resolver. But that should be done in a separate ticket.

Raised a new ticket for that: [#6379](https://github.com/twisted/twisted/issues/6379)

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_7"></a>|@jerub set owner to @wallrj|
|-|-|

I approve this change, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj set status to `closed`|
|-|-|

(In [37639]) Merge testable-resolver-3908: Pass the parameterized reactor to the DNSDatagramProtocol to allow easier testing.

Author: rwall
Reviewer: tom.prince, jerub
Fixes: [#3908](https://github.com/twisted/twisted/issues/3908)

Pass the parameterized reactor to the DNSDatagramProtocol to allow
easier testing.
